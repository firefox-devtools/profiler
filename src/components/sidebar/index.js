/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import * as React from 'react';

import { CallTreeSidebar } from './CallTreeSidebar';
import { MarkerSidebar } from './MarkerSidebar';

import type { TabSlug } from 'firefox-profiler/app-logic/tabs-handling';

import './sidebar.css';

export function selectSidebar(
  selectedTab: TabSlug
): React.ComponentType<{||}> | null {
  return {
    calltree: CallTreeSidebar,
    'function-list': CallTreeSidebar,
    'flame-graph': CallTreeSidebar,
    'stack-chart': null,
    'marker-chart': null,
    'marker-table': MarkerSidebar, // MarkerSidebar
    'network-chart': null,
    'js-tracer': null,
  }[selectedTab];
}
