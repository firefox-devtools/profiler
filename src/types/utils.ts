/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/**
 * This type serves as documentation for how an array is meant to be used, but does
 * not support type checking. We often use an Array instead of a Map to translate
 * one type of index into another type of index. This is similar to how we use the
 * Map<K,V> type, but with the Array.
 */
export type IndexedArray<_IndexType, Value> = Array<Value>;

export type MixedObject = { [key: string]: unknown };
