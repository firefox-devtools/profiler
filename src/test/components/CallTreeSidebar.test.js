/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import * as React from 'react';
import { Provider } from 'react-redux';

import { render, act } from 'firefox-profiler/test/fixtures/testing-library';
import { CallTreeSidebar } from '../../components/sidebar/CallTreeSidebar';
import {
  changeSelectedCallNode,
  changeInvertCallstack,
} from '../../actions/profile-view';

import { storeWithProfile } from '../fixtures/stores';
import { getProfileFromTextSamples } from '../fixtures/profiles/processed-profile';

import type { CallNodePath } from 'firefox-profiler/types';
import { ensureExists } from '../../utils/flow';
import { fireFullClick } from '../fixtures/utils';

describe('CallTreeSidebar', function () {
  function getProfileWithCategories() {
    return getProfileFromTextSamples(`
      A    A    A              A
      B    B    B              B
      Cjs  Cjs  H[cat:Layout]  H[cat:Layout]
      D    F    I[cat:Idle]
      Ejs  Ejs
    `);
  }

  function getProfileWithSubCategories() {
    const result = getProfileFromTextSamples(`
      A              A              A
      B              B              B
      C[cat:Layout]  C[cat:Layout]  C[cat:Layout]
                                    D[cat:Layout]
    `);

    const {
      profile,
      funcNamesDictPerThread: [{ C, D }],
    } = result;
    const layout = ensureExists(
      ensureExists(
        profile.meta.categories,
        'Expected to find categories.'
      ).find((category) => category.name === 'Layout'),
      'Could not find Layout category.'
    );
    const [{ frameTable }] = profile.threads;
    const fakeC = layout.subcategories.length;
    layout.subcategories.push('FakeSubCategoryC');
    const fakeD = layout.subcategories.length;
    layout.subcategories.push('FakeSubCategoryD');

    // The frames, funcs, and stacks all share the same indexes with the layout
    // of the stacks.
    frameTable.subcategory[C] = fakeC;
    frameTable.subcategory[D] = fakeD;

    return result;
  }

  function setup({ profile, funcNamesDictPerThread }) {
    const store = storeWithProfile(profile);

    const selectNode = (nodePath: CallNodePath) => {
      act(() => {
        store.dispatch(changeSelectedCallNode(0, nodePath));
      });
    };

    const invertCallstack = () =>
      act(() => store.dispatch(changeInvertCallstack(true)));

    const renderResult = render(
      <Provider store={store}>
        <CallTreeSidebar />
      </Provider>
    );

    const rerenderContainer = () =>
      renderResult.rerender(
        <Provider store={store}>
          <CallTreeSidebar />
        </Provider>
      );

    return {
      ...renderResult,
      rerenderContainer,
      ...store,
      funcNamesDict: funcNamesDictPerThread[0],
      selectNode,
      invertCallstack,
    };
  }

  it('matches the snapshots when displaying data about the currently selected node', () => {
    const {
      selectNode,
      funcNamesDict: { A, B, Cjs, D, H, Ejs },
      container,
    } = setup(getProfileWithCategories());

    expect(container.firstChild).toMatchSnapshot();

    // Cjs is a JS node, but has no self time, so we shouldn't see the
    // implementation information.
    selectNode([A, B, Cjs]);
    expect(container.firstChild).toMatchSnapshot();

    selectNode([A, B, Cjs, D]);
    expect(container.firstChild).toMatchSnapshot();

    selectNode([A, B, H]);
    expect(container.firstChild).toMatchSnapshot();

    selectNode([A, B, Cjs, D, Ejs]);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches the snapshots when displaying bytes data', () => {
    const profileWithDicts = getProfileWithCategories();
    // Create a weighted samples table with bytes.
    const [{ samples }] = profileWithDicts.profile.threads;
    samples.weightType = 'bytes';
    samples.weight = samples.stack.map((_stack, i) => i);

    const {
      selectNode,
      funcNamesDict: { A, B, Cjs },
      container,
    } = setup(profileWithDicts);

    expect(container.firstChild).toMatchSnapshot();

    selectNode([A, B, Cjs]);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches the snapshots when displaying data about the currently selected node in an inverted tree', () => {
    const {
      selectNode,
      invertCallstack,
      funcNamesDict: { A, B, H, Ejs, I },
      container,
    } = setup(getProfileWithCategories());

    invertCallstack();
    expect(container.firstChild).toMatchSnapshot();

    selectNode([Ejs]);
    expect(container.firstChild).toMatchSnapshot();

    selectNode([H]);
    expect(container.firstChild).toMatchSnapshot();

    selectNode([I, H]);
    expect(container.firstChild).toMatchSnapshot();

    selectNode([H, B, A]);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('can expand subcategories', () => {
    const {
      selectNode,
      container,
      queryByText,
      getAllByText,
      funcNamesDict: { A, B, C },
      rerenderContainer,
    } = setup(getProfileWithSubCategories());
    selectNode([A, B, C]);
    expect(queryByText('FakeSubCategoryC')).not.toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();

    const layoutCategory = getAllByText('Layout')[0];

    fireFullClick(layoutCategory);
    rerenderContainer();

    expect(getAllByText('FakeSubCategoryC')[0]).toBeInTheDocument();
    // only the 'Layout' category for the total running samples is expanded,
    // not the other one too
    expect(getAllByText('FakeSubCategoryC').length).toBe(1);

    expect(container.firstChild).toMatchSnapshot();

    const layoutCategory2 = getAllByText('Layout')[1];

    fireFullClick(layoutCategory2);
    rerenderContainer();

    expect(getAllByText('FakeSubCategoryC')[0]).toBeInTheDocument();
    expect(getAllByText('FakeSubCategoryC')[1]).toBeInTheDocument();
  });
});
