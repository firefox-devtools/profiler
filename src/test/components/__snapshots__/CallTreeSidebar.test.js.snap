// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CallTreeSidebar can expand subcategories 1`] = `
<aside
  class="sidebar sidebar-calltree"
>
  <div
    class="sidebar-contents-wrapper"
  >
    <header
      class="sidebar-titlegroup"
    >
      <h2
        class="sidebar-title can-select-content"
        title="C
(click to select)"
      >
        <input
          class="can-select-content-input"
          readonly=""
          value="C"
        />
      </h2>
    </header>
    <h4
      class="sidebar-title3"
    >
      <div>
        Call node details
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      Traced running time
    </div>
    <div
      class="sidebar-percentage"
    />
    <div
      class="sidebar-value"
    >
      3.0ms
    </div>
    <div
      class="sidebar-label"
    >
      Traced self time
    </div>
    <div
      class="sidebar-percentage"
    />
    <div
      class="sidebar-value"
    >
      2.0ms
    </div>
    <div
      class="sidebar-label"
    >
      Running samples
    </div>
    <div
      class="sidebar-percentage"
    >
      100%
    </div>
    <div
      class="sidebar-value"
    >
      3
    </div>
    <div
      class="sidebar-label"
    >
      Self samples
    </div>
    <div
      class="sidebar-percentage"
    >
      67%
    </div>
    <div
      class="sidebar-value"
    >
      2
    </div>
    <h4
      class="sidebar-title3 sidebar-title-label"
    >
      <div
        class="sidebar-title-label-left"
      >
        Categories
      </div>
      <div
        class="sidebar-title-label-right"
      >
        Running sample count
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      <button
        class="sidebar-toggle"
        data-category-index="2"
        type="button"
      >
        Layout
      </button>
    </div>
    <div
      class="sidebar-percentage"
    >
      100%
    </div>
    <div
      class="sidebar-value"
    >
      3
    </div>
    <div
      class="sidebar-histogram-bar"
    >
      <div
        class="sidebar-histogram-bar-color category-color-purple"
        style="width: 100.0000%;"
      />
    </div>
    <h4
      class="sidebar-title3 sidebar-title-label"
    >
      <div
        class="sidebar-title-label-left"
      >
        Categories
      </div>
      <div
        class="sidebar-title-label-right"
      >
        Self sample count
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      <button
        class="sidebar-toggle"
        data-category-index="2"
        type="button"
      >
        Layout
      </button>
    </div>
    <div
      class="sidebar-percentage"
    >
      100%
    </div>
    <div
      class="sidebar-value"
    >
      2
    </div>
    <div
      class="sidebar-histogram-bar"
    >
      <div
        class="sidebar-histogram-bar-color category-color-purple"
        style="width: 100.0000%;"
      />
    </div>
  </div>
</aside>
`;

exports[`CallTreeSidebar can expand subcategories 2`] = `
<aside
  class="sidebar sidebar-calltree"
>
  <div
    class="sidebar-contents-wrapper"
  >
    <header
      class="sidebar-titlegroup"
    >
      <h2
        class="sidebar-title can-select-content"
        title="C
(click to select)"
      >
        <input
          class="can-select-content-input"
          readonly=""
          value="C"
        />
      </h2>
    </header>
    <h4
      class="sidebar-title3"
    >
      <div>
        Call node details
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      Traced running time
    </div>
    <div
      class="sidebar-percentage"
    />
    <div
      class="sidebar-value"
    >
      3.0ms
    </div>
    <div
      class="sidebar-label"
    >
      Traced self time
    </div>
    <div
      class="sidebar-percentage"
    />
    <div
      class="sidebar-value"
    >
      2.0ms
    </div>
    <div
      class="sidebar-label"
    >
      Running samples
    </div>
    <div
      class="sidebar-percentage"
    >
      100%
    </div>
    <div
      class="sidebar-value"
    >
      3
    </div>
    <div
      class="sidebar-label"
    >
      Self samples
    </div>
    <div
      class="sidebar-percentage"
    >
      67%
    </div>
    <div
      class="sidebar-value"
    >
      2
    </div>
    <h4
      class="sidebar-title3 sidebar-title-label"
    >
      <div
        class="sidebar-title-label-left"
      >
        Categories
      </div>
      <div
        class="sidebar-title-label-right"
      >
        Running sample count
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      <button
        class="sidebar-toggle expanded"
        data-category-index="2"
        type="button"
      >
        Layout
      </button>
    </div>
    <div
      class="sidebar-percentage"
    >
      100%
    </div>
    <div
      class="sidebar-value"
    >
      3
    </div>
    <div
      class="sidebar-histogram-bar"
    >
      <div
        class="sidebar-histogram-bar-color category-color-purple"
        style="width: 100.0000%;"
      />
    </div>
    <div
      class="sidebar-label sidebar-label-indent"
    >
      FakeSubCategoryC
    </div>
    <div
      class="sidebar-percentage"
    >
      67%
    </div>
    <div
      class="sidebar-value"
    >
      2
    </div>
    <div
      class="sidebar-label sidebar-label-indent"
    >
      FakeSubCategoryD
    </div>
    <div
      class="sidebar-percentage"
    >
      33%
    </div>
    <div
      class="sidebar-value"
    >
      1
    </div>
    <h4
      class="sidebar-title3 sidebar-title-label"
    >
      <div
        class="sidebar-title-label-left"
      >
        Categories
      </div>
      <div
        class="sidebar-title-label-right"
      >
        Self sample count
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      <button
        class="sidebar-toggle"
        data-category-index="2"
        type="button"
      >
        Layout
      </button>
    </div>
    <div
      class="sidebar-percentage"
    >
      100%
    </div>
    <div
      class="sidebar-value"
    >
      2
    </div>
    <div
      class="sidebar-histogram-bar"
    >
      <div
        class="sidebar-histogram-bar-color category-color-purple"
        style="width: 100.0000%;"
      />
    </div>
  </div>
</aside>
`;

exports[`CallTreeSidebar matches the snapshots when displaying bytes data 1`] = `
<div
  class="sidebar sidebar-calltree"
>
  <div
    class="sidebar-contents-wrapper"
  >
    Select a node to display information about it.
  </div>
</div>
`;

exports[`CallTreeSidebar matches the snapshots when displaying bytes data 2`] = `
<aside
  class="sidebar sidebar-calltree"
>
  <div
    class="sidebar-contents-wrapper"
  >
    <header
      class="sidebar-titlegroup"
    >
      <h2
        class="sidebar-title can-select-content"
        title="Cjs
(click to select)"
      >
        <input
          class="can-select-content-input"
          readonly=""
          value="Cjs"
        />
      </h2>
    </header>
    <h4
      class="sidebar-title3"
    >
      <div>
        Call node details
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      Running size
    </div>
    <div
      class="sidebar-percentage"
    >
      17%
    </div>
    <div
      class="sidebar-value"
    >
      1B
    </div>
    <div
      class="sidebar-label"
    >
      Self size
    </div>
    <div
      class="sidebar-percentage"
    >
      —
    </div>
    <div
      class="sidebar-value"
    >
      —
    </div>
    <h4
      class="sidebar-title3 sidebar-title-label"
    >
      <div
        class="sidebar-title-label-left"
      >
        Categories
      </div>
      <div
        class="sidebar-title-label-right"
      >
        Running bytes
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      JavaScript
    </div>
    <div
      class="sidebar-percentage"
    >
      100%
    </div>
    <div
      class="sidebar-value"
    >
      1B
    </div>
    <div
      class="sidebar-histogram-bar"
    >
      <div
        class="sidebar-histogram-bar-color category-color-yellow"
        style="width: 100.0000%;"
      />
    </div>
  </div>
</aside>
`;

exports[`CallTreeSidebar matches the snapshots when displaying data about the currently selected node 1`] = `
<div
  class="sidebar sidebar-calltree"
>
  <div
    class="sidebar-contents-wrapper"
  >
    Select a node to display information about it.
  </div>
</div>
`;

exports[`CallTreeSidebar matches the snapshots when displaying data about the currently selected node 2`] = `
<aside
  class="sidebar sidebar-calltree"
>
  <div
    class="sidebar-contents-wrapper"
  >
    <header
      class="sidebar-titlegroup"
    >
      <h2
        class="sidebar-title can-select-content"
        title="Cjs
(click to select)"
      >
        <input
          class="can-select-content-input"
          readonly=""
          value="Cjs"
        />
      </h2>
    </header>
    <h4
      class="sidebar-title3"
    >
      <div>
        Call node details
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      Traced running time
    </div>
    <div
      class="sidebar-percentage"
    />
    <div
      class="sidebar-value"
    >
      2.0ms
    </div>
    <div
      class="sidebar-label"
    >
      Traced self time
    </div>
    <div
      class="sidebar-percentage"
    />
    <div
      class="sidebar-value"
    >
      —
    </div>
    <div
      class="sidebar-label"
    >
      Running samples
    </div>
    <div
      class="sidebar-percentage"
    >
      50%
    </div>
    <div
      class="sidebar-value"
    >
      2
    </div>
    <div
      class="sidebar-label"
    >
      Self samples
    </div>
    <div
      class="sidebar-percentage"
    >
      —
    </div>
    <div
      class="sidebar-value"
    >
      —
    </div>
    <h4
      class="sidebar-title3 sidebar-title-label"
    >
      <div
        class="sidebar-title-label-left"
      >
        Categories
      </div>
      <div
        class="sidebar-title-label-right"
      >
        Running sample count
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      JavaScript
    </div>
    <div
      class="sidebar-percentage"
    >
      100%
    </div>
    <div
      class="sidebar-value"
    >
      2
    </div>
    <div
      class="sidebar-histogram-bar"
    >
      <div
        class="sidebar-histogram-bar-color category-color-yellow"
        style="width: 100.0000%;"
      />
    </div>
  </div>
</aside>
`;

exports[`CallTreeSidebar matches the snapshots when displaying data about the currently selected node 3`] = `
<aside
  class="sidebar sidebar-calltree"
>
  <div
    class="sidebar-contents-wrapper"
  >
    <header
      class="sidebar-titlegroup"
    >
      <h2
        class="sidebar-title can-select-content"
        title="D
(click to select)"
      >
        <input
          class="can-select-content-input"
          readonly=""
          value="D"
        />
      </h2>
    </header>
    <h4
      class="sidebar-title3"
    >
      <div>
        Call node details
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      Traced running time
    </div>
    <div
      class="sidebar-percentage"
    />
    <div
      class="sidebar-value"
    >
      1.0ms
    </div>
    <div
      class="sidebar-label"
    >
      Traced self time
    </div>
    <div
      class="sidebar-percentage"
    />
    <div
      class="sidebar-value"
    >
      —
    </div>
    <div
      class="sidebar-label"
    >
      Running samples
    </div>
    <div
      class="sidebar-percentage"
    >
      25%
    </div>
    <div
      class="sidebar-value"
    >
      1
    </div>
    <div
      class="sidebar-label"
    >
      Self samples
    </div>
    <div
      class="sidebar-percentage"
    >
      —
    </div>
    <div
      class="sidebar-value"
    >
      —
    </div>
    <h4
      class="sidebar-title3 sidebar-title-label"
    >
      <div
        class="sidebar-title-label-left"
      >
        Categories
      </div>
      <div
        class="sidebar-title-label-right"
      >
        Running sample count
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      JavaScript
    </div>
    <div
      class="sidebar-percentage"
    >
      100%
    </div>
    <div
      class="sidebar-value"
    >
      1
    </div>
    <div
      class="sidebar-histogram-bar"
    >
      <div
        class="sidebar-histogram-bar-color category-color-yellow"
        style="width: 100.0000%;"
      />
    </div>
  </div>
</aside>
`;

exports[`CallTreeSidebar matches the snapshots when displaying data about the currently selected node 4`] = `
<aside
  class="sidebar sidebar-calltree"
>
  <div
    class="sidebar-contents-wrapper"
  >
    <header
      class="sidebar-titlegroup"
    >
      <h2
        class="sidebar-title can-select-content"
        title="H
(click to select)"
      >
        <input
          class="can-select-content-input"
          readonly=""
          value="H"
        />
      </h2>
    </header>
    <h4
      class="sidebar-title3"
    >
      <div>
        Call node details
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      Traced running time
    </div>
    <div
      class="sidebar-percentage"
    />
    <div
      class="sidebar-value"
    >
      2.0ms
    </div>
    <div
      class="sidebar-label"
    >
      Traced self time
    </div>
    <div
      class="sidebar-percentage"
    />
    <div
      class="sidebar-value"
    >
      1.0ms
    </div>
    <div
      class="sidebar-label"
    >
      Running samples
    </div>
    <div
      class="sidebar-percentage"
    >
      50%
    </div>
    <div
      class="sidebar-value"
    >
      2
    </div>
    <div
      class="sidebar-label"
    >
      Self samples
    </div>
    <div
      class="sidebar-percentage"
    >
      25%
    </div>
    <div
      class="sidebar-value"
    >
      1
    </div>
    <h4
      class="sidebar-title3 sidebar-title-label"
    >
      <div
        class="sidebar-title-label-left"
      >
        Categories
      </div>
      <div
        class="sidebar-title-label-right"
      >
        Running sample count
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      Idle
    </div>
    <div
      class="sidebar-percentage"
    >
      50%
    </div>
    <div
      class="sidebar-value"
    >
      1
    </div>
    <div
      class="sidebar-histogram-bar"
    >
      <div
        class="sidebar-histogram-bar-color category-color-transparent"
        style="width: 50.0000%;"
      />
    </div>
    <div
      class="sidebar-label"
    >
      Layout
    </div>
    <div
      class="sidebar-percentage"
    >
      50%
    </div>
    <div
      class="sidebar-value"
    >
      1
    </div>
    <div
      class="sidebar-histogram-bar"
    >
      <div
        class="sidebar-histogram-bar-color category-color-purple"
        style="width: 50.0000%;"
      />
    </div>
    <h4
      class="sidebar-title3 sidebar-title-label"
    >
      <div
        class="sidebar-title-label-left"
      >
        Categories
      </div>
      <div
        class="sidebar-title-label-right"
      >
        Self sample count
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      Layout
    </div>
    <div
      class="sidebar-percentage"
    >
      100%
    </div>
    <div
      class="sidebar-value"
    >
      1
    </div>
    <div
      class="sidebar-histogram-bar"
    >
      <div
        class="sidebar-histogram-bar-color category-color-purple"
        style="width: 100.0000%;"
      />
    </div>
  </div>
</aside>
`;

exports[`CallTreeSidebar matches the snapshots when displaying data about the currently selected node 5`] = `
<aside
  class="sidebar sidebar-calltree"
>
  <div
    class="sidebar-contents-wrapper"
  >
    <header
      class="sidebar-titlegroup"
    >
      <h2
        class="sidebar-title can-select-content"
        title="Ejs
(click to select)"
      >
        <input
          class="can-select-content-input"
          readonly=""
          value="Ejs"
        />
      </h2>
    </header>
    <h4
      class="sidebar-title3"
    >
      <div>
        Call node details
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      Traced running time
    </div>
    <div
      class="sidebar-percentage"
    />
    <div
      class="sidebar-value"
    >
      1.0ms
    </div>
    <div
      class="sidebar-label"
    >
      Traced self time
    </div>
    <div
      class="sidebar-percentage"
    />
    <div
      class="sidebar-value"
    >
      1.0ms
    </div>
    <div
      class="sidebar-label"
    >
      Running samples
    </div>
    <div
      class="sidebar-percentage"
    >
      25%
    </div>
    <div
      class="sidebar-value"
    >
      1
    </div>
    <div
      class="sidebar-label"
    >
      Self samples
    </div>
    <div
      class="sidebar-percentage"
    >
      25%
    </div>
    <div
      class="sidebar-value"
    >
      1
    </div>
    <h4
      class="sidebar-title3 sidebar-title-label"
    >
      <div
        class="sidebar-title-label-left"
      >
        Categories
      </div>
      <div
        class="sidebar-title-label-right"
      >
        Running sample count
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      JavaScript
    </div>
    <div
      class="sidebar-percentage"
    >
      100%
    </div>
    <div
      class="sidebar-value"
    >
      1
    </div>
    <div
      class="sidebar-histogram-bar"
    >
      <div
        class="sidebar-histogram-bar-color category-color-yellow"
        style="width: 100.0000%;"
      />
    </div>
    <h4
      class="sidebar-title3 sidebar-title-label"
    >
      <div
        class="sidebar-title-label-left"
      >
        Categories
      </div>
      <div
        class="sidebar-title-label-right"
      >
        Self sample count
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      JavaScript
    </div>
    <div
      class="sidebar-percentage"
    >
      100%
    </div>
    <div
      class="sidebar-value"
    >
      1
    </div>
    <div
      class="sidebar-histogram-bar"
    >
      <div
        class="sidebar-histogram-bar-color category-color-yellow"
        style="width: 100.0000%;"
      />
    </div>
  </div>
</aside>
`;

exports[`CallTreeSidebar matches the snapshots when displaying data about the currently selected node in an inverted tree 1`] = `
<div
  class="sidebar sidebar-calltree"
>
  <div
    class="sidebar-contents-wrapper"
  >
    Select a node to display information about it.
  </div>
</div>
`;

exports[`CallTreeSidebar matches the snapshots when displaying data about the currently selected node in an inverted tree 2`] = `
<aside
  class="sidebar sidebar-calltree"
>
  <div
    class="sidebar-contents-wrapper"
  >
    <header
      class="sidebar-titlegroup"
    >
      <h2
        class="sidebar-title can-select-content"
        title="Ejs
(click to select)"
      >
        <input
          class="can-select-content-input"
          readonly=""
          value="Ejs"
        />
      </h2>
    </header>
    <h4
      class="sidebar-title3"
    >
      <div>
        Call node details
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      Traced running time
    </div>
    <div
      class="sidebar-percentage"
    />
    <div
      class="sidebar-value"
    >
      2.0ms
    </div>
    <div
      class="sidebar-label"
    >
      Traced self time
    </div>
    <div
      class="sidebar-percentage"
    />
    <div
      class="sidebar-value"
    >
      2.0ms
    </div>
    <div
      class="sidebar-label"
    >
      Running samples
    </div>
    <div
      class="sidebar-percentage"
    >
      50%
    </div>
    <div
      class="sidebar-value"
    >
      2
    </div>
    <div
      class="sidebar-label"
    >
      Self samples
    </div>
    <div
      class="sidebar-percentage"
    >
      50%
    </div>
    <div
      class="sidebar-value"
    >
      2
    </div>
    <h4
      class="sidebar-title3 sidebar-title-label"
    >
      <div
        class="sidebar-title-label-left"
      >
        Categories
      </div>
      <div
        class="sidebar-title-label-right"
      >
        Running sample count
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      JavaScript
    </div>
    <div
      class="sidebar-percentage"
    >
      100%
    </div>
    <div
      class="sidebar-value"
    >
      2
    </div>
    <div
      class="sidebar-histogram-bar"
    >
      <div
        class="sidebar-histogram-bar-color category-color-yellow"
        style="width: 100.0000%;"
      />
    </div>
  </div>
</aside>
`;

exports[`CallTreeSidebar matches the snapshots when displaying data about the currently selected node in an inverted tree 3`] = `
<aside
  class="sidebar sidebar-calltree"
>
  <div
    class="sidebar-contents-wrapper"
  >
    <header
      class="sidebar-titlegroup"
    >
      <h2
        class="sidebar-title can-select-content"
        title="H
(click to select)"
      >
        <input
          class="can-select-content-input"
          readonly=""
          value="H"
        />
      </h2>
    </header>
    <h4
      class="sidebar-title3"
    >
      <div>
        Call node details
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      Traced running time
    </div>
    <div
      class="sidebar-percentage"
    />
    <div
      class="sidebar-value"
    >
      1.0ms
    </div>
    <div
      class="sidebar-label"
    >
      Traced self time
    </div>
    <div
      class="sidebar-percentage"
    />
    <div
      class="sidebar-value"
    >
      1.0ms
    </div>
    <div
      class="sidebar-label"
    >
      Running samples
    </div>
    <div
      class="sidebar-percentage"
    >
      25%
    </div>
    <div
      class="sidebar-value"
    >
      1
    </div>
    <div
      class="sidebar-label"
    >
      Self samples
    </div>
    <div
      class="sidebar-percentage"
    >
      25%
    </div>
    <div
      class="sidebar-value"
    >
      1
    </div>
    <h4
      class="sidebar-title3 sidebar-title-label"
    >
      <div
        class="sidebar-title-label-left"
      >
        Categories
      </div>
      <div
        class="sidebar-title-label-right"
      >
        Running sample count
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      Layout
    </div>
    <div
      class="sidebar-percentage"
    >
      100%
    </div>
    <div
      class="sidebar-value"
    >
      1
    </div>
    <div
      class="sidebar-histogram-bar"
    >
      <div
        class="sidebar-histogram-bar-color category-color-purple"
        style="width: 100.0000%;"
      />
    </div>
  </div>
</aside>
`;

exports[`CallTreeSidebar matches the snapshots when displaying data about the currently selected node in an inverted tree 4`] = `
<aside
  class="sidebar sidebar-calltree"
>
  <div
    class="sidebar-contents-wrapper"
  >
    <header
      class="sidebar-titlegroup"
    >
      <h2
        class="sidebar-title can-select-content"
        title="H
(click to select)"
      >
        <input
          class="can-select-content-input"
          readonly=""
          value="H"
        />
      </h2>
    </header>
    <h4
      class="sidebar-title3"
    >
      <div>
        Call node details
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      Traced running time
    </div>
    <div
      class="sidebar-percentage"
    />
    <div
      class="sidebar-value"
    >
      1.0ms
    </div>
    <div
      class="sidebar-label"
    >
      Traced self time
    </div>
    <div
      class="sidebar-percentage"
    />
    <div
      class="sidebar-value"
    >
      —
    </div>
    <div
      class="sidebar-label"
    >
      Running samples
    </div>
    <div
      class="sidebar-percentage"
    >
      25%
    </div>
    <div
      class="sidebar-value"
    >
      1
    </div>
    <div
      class="sidebar-label"
    >
      Self samples
    </div>
    <div
      class="sidebar-percentage"
    >
      —
    </div>
    <div
      class="sidebar-value"
    >
      —
    </div>
    <h4
      class="sidebar-title3 sidebar-title-label"
    >
      <div
        class="sidebar-title-label-left"
      >
        Categories
      </div>
      <div
        class="sidebar-title-label-right"
      >
        Running sample count
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      Idle
    </div>
    <div
      class="sidebar-percentage"
    >
      100%
    </div>
    <div
      class="sidebar-value"
    >
      1
    </div>
    <div
      class="sidebar-histogram-bar"
    >
      <div
        class="sidebar-histogram-bar-color category-color-transparent"
        style="width: 100.0000%;"
      />
    </div>
  </div>
</aside>
`;

exports[`CallTreeSidebar matches the snapshots when displaying data about the currently selected node in an inverted tree 5`] = `
<aside
  class="sidebar sidebar-calltree"
>
  <div
    class="sidebar-contents-wrapper"
  >
    <header
      class="sidebar-titlegroup"
    >
      <h2
        class="sidebar-title can-select-content"
        title="A
(click to select)"
      >
        <input
          class="can-select-content-input"
          readonly=""
          value="A"
        />
      </h2>
    </header>
    <h4
      class="sidebar-title3"
    >
      <div>
        Call node details
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      Traced running time
    </div>
    <div
      class="sidebar-percentage"
    />
    <div
      class="sidebar-value"
    >
      1.0ms
    </div>
    <div
      class="sidebar-label"
    >
      Traced self time
    </div>
    <div
      class="sidebar-percentage"
    />
    <div
      class="sidebar-value"
    >
      —
    </div>
    <div
      class="sidebar-label"
    >
      Running samples
    </div>
    <div
      class="sidebar-percentage"
    >
      25%
    </div>
    <div
      class="sidebar-value"
    >
      1
    </div>
    <div
      class="sidebar-label"
    >
      Self samples
    </div>
    <div
      class="sidebar-percentage"
    >
      —
    </div>
    <div
      class="sidebar-value"
    >
      —
    </div>
    <h4
      class="sidebar-title3 sidebar-title-label"
    >
      <div
        class="sidebar-title-label-left"
      >
        Categories
      </div>
      <div
        class="sidebar-title-label-right"
      >
        Running sample count
      </div>
    </h4>
    <div
      class="sidebar-label"
    >
      Layout
    </div>
    <div
      class="sidebar-percentage"
    >
      100%
    </div>
    <div
      class="sidebar-value"
    >
      1
    </div>
    <div
      class="sidebar-histogram-bar"
    >
      <div
        class="sidebar-histogram-bar-color category-color-purple"
        style="width: 100.0000%;"
      />
    </div>
  </div>
</aside>
`;
